### PR TITLE
go/worker/p2p: Fix message relaying and retrying on errors

### PR DESCRIPTION
**Pull request by ptrus**

### CODE REVIEW

**Review by Yawning — APPROVED** [review]

**Review by kostko — APPROVED** [review]

File: go/worker/common/p2p/error/error.go [hunk 1]
```diff
@@ -23,6 +57,40 @@ func IsPermanent(err error) bool {
...
+}
+
+// EnsurePermanent ensures an error will be correctly treated as (non-)permanent
+// by `cenkalti/backoff/v4`.
+//
+// Note: `IsPermanent` notion of a permanent error differs from the `cenkalti/backoff/v4`:
+// - it special cases `context.Canceled`
+// - it correctly handles wrapped permanent errors
+func EnsurePermanent(err error) error {
+	// TODO: file an issue upstream to make use of `errors.As`?
```
Comment (kostko) [inline go/worker/common/p2p/error/error.go:71]: Yeah that is a good idea, can you do it?

Comment (ptrus) [reply go/worker/common/p2p/error/error.go:71]: yeah can do

Comment (kostko) [reply go/worker/common/p2p/error/error.go:71]: Thanks!

Comment (ptrus) [reply go/worker/common/p2p/error/error.go]: done https://github.com/cenkalti/backoff/issues/107

**Pull request merged**